knip.jsonc

```diff
@@ -0,0 +1,46 @@
+{
+    "workspaces": {
+        ".": {
+            // These entries are complementary to the ones found in package.json
+            "entry": [
+                "lib/rules/index.js",
```

Comment:
[Question] Is there a tracking issue somewhere for getting Knip to auto-detect this? I don't think ESLint is doing anything particularly spooky.

`package.json`'s `"bin"` -> `bin/eslint.js` -> `lib/cli` (`cli.execute`) -> `lib/eslint/eslint.js` (`ESLint`) -> `lib/linter/linter.js` (`Linter`) -> `lib/linter/rules` (`Rules`) -> `lib/rules` (`builtInRules`)

Reply:
I've briefly touched upon this in the pull request description. Let me elaborate a bit more. CommonJS allows some freedom in using `module.exports`, the following is all valid:

```js
module.exports = function exportedFunction() {}; // 1
```

```js
module.exports = { functionA, functionB }; // 2
```

```js
module.exports.functionA = functionA; // 3
module.exports.functionB = functionB;
```

For (1) it's unambiguous that this is the default export. For (2) however, the object could be considered the default object, or perhaps the _intention_ is basically what (3) does. All of this is valid and reasonable within the context of CommonJS, no matter how things are imported.

However, for Knip (or any static analysis tool in this field for that matter) to give meaningful results about what exports are being used or not, a strategy has to be decided upon. We can't say "this is a default export AND named exports" at the same time without getting false positives. It's also valid to default-export an object that does not represent named exports, so their members are not meant to be imported by name:

```js
module.exports = { option: true, path: '/' };
```

And then import this:

```js
import Config from './config'; // common
import { option } from './config'; // valid, but not common
```


I've found that in the majority of cases, the intention of example (2) is to do assume that those are named exports, exactly like (3) does. And similarly, those named exports of (2) and (3) are imported using named imports as well.

Overall, this strategy works relatively well. But as we can see in this PR, some modifications to source code are to be made for Knip in order to give correct results in this area of "unused exports". In case the flexibility of CommonJS is appreciated and there's hesitance towards changing source code for the sake of a linter, that's totally reasonable. On the other hand, following the same strategy can lead to more consistent code and an easier migration path if ES Modules would become a goal.

The nice thing about ES Modules is that it does not have this problem of ambiguity.

Going back to the question, "the chain breaks" (for Knip) at `require("../lib/cli").execute` in `bin/eslint.js` where the named `execute` export is used, while `lib/cli.js` has a default export only.

Reply:
I filed a bug report here: https://github.com/webpro/knip/issues/465. The explanation in https://github.com/webpro/knip/issues/465#issuecomment-1901889825 gives more info too (though re-reading this, most of the info was already here heh).

Reply:
Wrote a little guide about this here: https://knip.dev/guides/commonjs